stl/inc/xstring

```diff
@@ -2468,6 +2468,69 @@ public:
         _Proxy._Release();
     }
 
+    basic_string(_String_constructor_concat_tag, basic_string& _Left, basic_string& _Right)
+        : _Mypair(_One_then_variadic_args_t(), _Left._Getal()) {
+        auto& _My_data    = _Mypair._Myval2;
+        auto& _Left_data  = _Left._Mypair._Myval2;
+        auto& _Right_data = _Right._Mypair._Myval2;
+        _Left_data._Orphan_all();
+        _Right_data._Orphan_all();
+        const auto _Left_size  = _Left_data._Mysize;
+        const auto _Right_size = _Right_data._Mysize;
+
+        const auto _Left_capacity  = _Left_data._Myres;
+        const auto _Right_capacity = _Right_data._Myres;
+        // overflow is OK due to max_size() checks:
+        const auto _New_size = static_cast<size_type>(_Left_size + _Right_size);
+        if (_Allocators_equal(_Getal(), _Right._Getal()) && _Right_capacity > _Left_capacity
+            && _Left_size <= _Right_capacity - _Right_size) {
+            // take _Right's buffer, max_size() is OK because we're under _Right's capacity
+            _My_data._Alloc_proxy(_GET_PROXY_ALLOCATOR(_Alty, _Getal())); // throws, hereafter nothrow in this block
+            _Take_contents(_Right, bool_constant<_Can_memcpy_val>{});
+            // _Right had more capacity than left, therefore _Right did not have the minimum capacity,
+            // therefore _Right had _Large_string_engaged(), therefore after _Take_contents we have
+            // _Large_string_engaged() and can save that branch
+            _STL_INTERNAL_CHECK(_My_data._Large_string_engaged());
+            const auto _Ptr = _Unfancy(_My_data._Bx._Ptr);
+            _Traits::move(_Ptr + _Left_size, _Ptr, _My_data._Mysize + 1);
+            _Traits::copy(_Ptr, _Left_data._Myptr(), _Left_size);
+            _My_data._Mysize = _New_size;
+            _Traits::assign(_Ptr[_New_size], _Elem());
+            return;
+        }
+
+        if (_Right_size < _Left_capacity - _Left_size) {
+            // take _Left's buffer, max_size() is OK because we're under _Left's capacity
+            _My_data._Alloc_proxy(_GET_PROXY_ALLOCATOR(_Alty, _Getal())); // throws, hereafter nothrow in this block
+            _Take_contents(_Left, bool_constant<_Can_memcpy_val>{});
+            const auto _Ptr = _My_data._Myptr();
+            _Traits::copy(_Ptr + _My_data._Mysize, _Right_data._Myptr(), _Right_size);

```

Comment:
If we +1, then we pay 1 instruction to do the increment `inc register`, then memcpy pays at least 1/32nd of an instruction to read that last byte, and 1/32nd of an instruction to write that byte (assuming AVX256 hardware).
If we explicitly assign, we pay 1 instruction to write the null. `mov BYTE_PTR[register+register], 0`
So I think overall the assign call is the better thing.

Reply:
*With hands waving* My mental performance model is based on number of memory accesses, rather than number of instructions. 2 * 1/32 == 1/16 of a memory access is less than 1.

Reply:
Fun whinging aside: It was more important to me that we `+1` or `assign` consistently in the two cases than that we pick any particular implementation.